src/include/miopen/conv/heur/memref.hpp

```diff
@@ -0,0 +1,97 @@
+/*******************************************************************************
+*
+* MIT License
+*
+* Copyright (c) 2020 Advanced Micro Devices, Inc.
+*
+* Permission is hereby granted, free of charge, to any person obtaining a copy
+* of this software and associated documentation files (the "Software"), to deal
+* in the Software without restriction, including without limitation the rights
+* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+* copies of the Software, and to permit persons to whom the Software is
+* furnished to do so, subject to the following conditions:
+*
+* The above copyright notice and this permission notice shall be included in all
+* copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+* SOFTWARE.
+*
+*******************************************************************************/
+
+#ifndef GUARD_MIOPEN_MEMREF_HPP_
+#define GUARD_MIOPEN_MEMREF_HPP_
+
+#include <stddef.h>
+#include <vector>
+#include <cstdint>
+#include <iostream>
+#include <cstring>
+#include <stdlib.h>
+#include <cassert>
+namespace miopen
+{
+
+struct MemRef2D
+{
+    float* aligned;
+    float* unaligned;
+    int64_t offset;
+    int64_t size0;
+    int64_t size1;
+    int64_t stride0;
+    int64_t stride1;
+    void print()
+    {
+        assert(aligned == unaligned);
+        for(auto idx0 = 0; idx0 < size0; ++idx0)
+        {
+            for(auto idx1 = 0; idx1 < size1; ++idx1)
+            {
+                std::cout << aligned[idx0 * stride0 + idx1 * stride1] << '\t';
+            }
+            std::cout << std::endl;
+        }
+    }
+};
```

Comment:
I was randomly browsing PRs in MIOpen and this memref catched my eye. The interaction with mlir will soon require me to also come up with a memref type so I'd hope we can together come up with one robust solution that can be used in shared scenarios.

Clarification question: Does this memref interface with external dependencies?

The one that is defined in mlir is [here](https://github.com/llvm/llvm-project/blob/master/mlir/include/mlir/ExecutionEngine/CRunnerUtils.h#L115-L155). That memref is almost similar to yours except that the first field and second field has swapped their positions: In MLIR, first field is unaligned. Although in your use cases, I always see the two fields getting similar values but I want to make sure they do meant to swap.

Is it possible to use the templatized version for extendability into different dimensions? Attaching the MLIR definition below:

```cpp
template <typename T, int N>
struct StridedMemRefType {
  T *basePtr;
  T *data;
  int64_t offset;
  int64_t sizes[N];
  int64_t strides[N];
  // This operator[] is extremely slow and only for sugaring purposes.
  StridedMemRefType<T, N - 1> operator[](int64_t idx) {
    StridedMemRefType<T, N - 1> res;
    res.basePtr = basePtr;
    res.data = data;
    res.offset = offset + idx * strides[0];
    dropFront<N>(sizes, res.sizes);
    dropFront<N>(strides, res.strides);
    return res;
  }
};

```

Reply:
@jerryyin  Thank you for pointing this out, I was not aware of that implementation as well as the bug you pointed out. I have removed my implementation and replaced it with the one in LLVM. Of course I have copied it over since I don't wish MIOpen to depend on LLVM for a single struct. 

I would appreciate if you can take a second look.

Reply:
The new changes that pull in MemRef2D looks good!

I don't think I understand all use case of MemRef2D in your PR so leave this for you to decide: In MLIR, Memref2d has exposed a lot of the internals of the data layout. When this needs to be exposed on a binary interface, an unranked memref which exposed fewer implementation details are typically used. If the Memref is used across binary interface, I'd suggest you to use unranked memref in that interface. If not, you are all good.